### DISCUSSION TITLE
Extension development (again)

### DISCUSSION THREAD

**Discussion by dasantonym** [directus/directus#19091] [Feature Requests]
### Summary

I know there are several posts dealing with the question of how to properly develop extensions, but I wanted to ask again for the current state of things.

I was developing an endpoint extension with several NPM dependencies and that was working for me in a locally running Directus 9.x repo. I used the SDK and produced a minified API extension that I copied to the extensions/endpoints folder. I was using the 10.0.0 SDK and I was copying the `dev` version of `index.js` (unminified).

I was debugging my extension using WebStorm attaching to `node dist/start.js` with the work dir `api` and I could place breakpoints in my extension's minified `index.js`. Not perfect, but everything was going well.

Now my shiny extension was done, so I used `build`, deployed and kaboom, syntax errors complaining about `)` characters and other weird things (unable to destructure `exceptions`).

Now things went further south as I upgraded my local repo to the current 10.4.2 version. Now I cannot set a breakpoint, while breaks on exceptions sometimes triggers a stop here and there. Just not in my extension.

I also added the extension using `pnpm install ../../my-extension-project-folder`, but it was the same thing. I also placed my breakpoints on the dev-"minified" file under `dist/` in `node_modules`.

Then I realised I do not need the minification so I changed the `package.json` to use:

```json
	"directus:extension": {
		"type": "endpoint",
		"path": "src/index.js",
		"source": "src/index.js",
		"host": "^10.1.5"
	},
```

Now it seems to be unable to resolve the `import` statements pointing to NPM modules within the extension.

**TL;DR:**

- What is the current best practice for debugging extensions during development?
- How do I properly install an API extension with extra dependencies?

### Motivation

I'd like to use Directus more in my projects.

### Detailed Design

Well, I somehow think I should probably have opened an issue instead?

### Requirements List

Extensions should be less frustrating to develop.

### Drawbacks

...

### Alternatives

...

### Adoption Strategy

...

**Comment by dasantonym** [comment]
Minor addition: I also tried to adapt the settings proposed here, but no luck: https://docs.directus.io/contributing/running-locally.html#debugging-the-api-in-vs-code

**Comment by rijkvanzanten** [comment]
Closing this as it's not a feature request. Happy to discuss extensions on [Discord](https://directus.chat) 👍🏻

**Reply by dasantonym** [reply]
Yeah, sorry, I realised too late that this was totally off topic and did not want to discard the text... Anyway, Discord is not so much my cup of tea, but I'll remember this for next time!

Should anyone ever stumble on this, I did not get breakpoints in my endpoint extension under `api/node_modules`, because `extensions.ts` always added a `t=${Date.now()}` to the end of the import:

https://github.com/directus/directus/blob/55f3d4c6e0805fd62ad7203e109461a7fc3152c1/api/src/extensions.ts#L452-L454

Once I removed that, the breakpoints were hit again. Guess there is some more correct or elegant way to get around this (without editing `extensions.ts`, but it works for now.
